src/main/java/de/tum/in/www1/artemis/service/scheduled/ScheduleService.java

```diff
@@ -43,22 +46,23 @@ private void removeScheduledTask(Exercise exercise) {
     void scheduleTask(Exercise exercise, ExerciseLifecycle lifecycle, Runnable task) {
         // check if already scheduled for exercise. if so, cancel.
         // no exercise should be scheduled for clustering more than once.
-        cancelScheduledTask(exercise);
+        cancelScheduledTaskForLifecycle(exercise, lifecycle);
         ScheduledFuture scheduledTask = exerciseLifecycleService.scheduleTask(exercise, lifecycle, task);
-        addScheduledTask(exercise, scheduledTask);
+        addScheduledTask(exercise, lifecycle, scheduledTask);
     }
 
     /**
      * Cancel possible schedules tasks for a provided exercise.
      * @param exercise exercise for which a potential clustering task is canceled
      */
-    void cancelScheduledTask(Exercise exercise) {
-        ScheduledFuture future = scheduledTasks.get(exercise.getId());
+    void cancelScheduledTaskForLifecycle(Exercise exercise, ExerciseLifecycle lifecycle) {
+        Tuple<Long, ExerciseLifecycle> taskId = new Tuple<>(exercise.getId(), lifecycle);
+        ScheduledFuture future = scheduledTasks.get(taskId);
         if (future != null) {
             log.debug("Cancelling scheduled task for build and test for student submissions after due date for Programming Exercise \"" + exercise.getTitle() + "\" (#"
```

Comment:
I see that we introduced this in another PR, but I do not think a debug statement in the generic `ScheduleService` should mention details on a "Programming Exercise".

Reply:
I agree